### PR TITLE
add SU4 gate, update pytket

### PR DESCRIPTION
**Pull request by cqc-melf**

### CODE REVIEW

File: tests/qir/test_pytket_qir_quantum_v.ll [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  call void @__quantum__qis__h__body(%Qubit* null)
+  call void @__quantum__qis__x__body(%Qubit* null)
+  call void @__quantum__qis__y__body(%Qubit* null)
+  call void @__quantum__qis__z__body(%Qubit* null)
+  call void @__quantum__qis__rx__body(double 5.000000e-01, %Qubit* null)
+  call void @__quantum__qis__rzz__body(double 5.000000e-01, %Qubit* null, %Qubit* inttoptr (i64 1 to %Qubit*))
+  call void @__quantum__qis__phasedx__body(double 5.000000e-01, double 4.000000e-01, %Qubit* inttoptr (i64 1 to %Qubit*))
+  call void @__quantum__qis__rxxyyzz__body(double 0x3FF921FB54442D18, double 0x3FF41B2F769CF0E0, double 0x3FEE28C731EB6950, %Qubit* null, %Qubit* inttoptr (i64 1 to %Qubit*))
+  call void @__quantum__qis__rxxyyzz__body(double 0x4012D97C7F3321D2, double 0x401197C987C952C4, double 0x40105616905F83B6, %Qubit* inttoptr (i64 2 to %Qubit*), %Qubit* inttoptr (i64 1 to %Qubit*))
```
Comment (cqc-melf) [inline tests/qir/test_pytket_qir_quantum_v.ll:20]: I am a bit surprised by this float to string conversion in the file, but it seams to be working right.

File: setup.py [hunk 2]
```diff
@@ -43,11 +43,12 @@
     packages=find_namespace_packages(include=["pytket.*"]),
     include_package_data=True,
     install_requires=[
-        "pytket ~= 1.15",
+        "pytket ~= 1.16",
         "pyqir == 0.8.2",
         "pyqir-generator == 0.7.0",
         "pyqir-evaluator == 0.7.0",
         "pyqir-parser == 0.7.0",
+        "numpy",
```
Comment (cqc-alec) [inline setup.py]: Since we're only using this for `numpy.pi`, why don't we just use `math.pi` and avoid the extra dependency?

**Review by cqc-alec — CHANGES_REQUESTED** [review]
I don't understand why we are introducing this radians-conversion logic here when it should already exist for other parametrized rotation gates. Is it a bug that we are not converting those?

**Comment by cqc-melf** [comment]
> I don't understand why we are introducing this radians-conversion logic here when it should already exist for other parametrized rotation gates. Is it a bug that we are not converting those?

I have not added any parameter conversion for the gates. I will add them now. Is this something that is defined in the QIR profile or is that back end specific?

**Comment by cqc-alec** [comment]
> > I don't understand why we are introducing this radians-conversion logic here when it should already exist for other parametrized rotation gates. Is it a bug that we are not converting those?
> 
> I have not added any parameter conversion for the gates. I will add them now. Is this something that is defined in the QIR profile or is that back end specific?

I can't find chapter and verse, but am fairly certain it's general. As far as I know pytket is the only software that uses half-turn units for rotation angles.

**Comment by cqc-melf** [comment]
> > > I don't understand why we are introducing this radians-conversion logic here when it should already exist for other parametrized rotation gates. Is it a bug that we are not converting those?
> > 
> > 
> > I have not added any parameter conversion for the gates. I will add them now. Is this something that is defined in the QIR profile or is that back end specific?
> 
> I can't find chapter and verse, but am fairly certain it's general. As far as I know pytket is the only software that uses half-turn units for rotation angles.

Sounds plausible, I was thinking if we should add a parameter for this in the API, but I have not done this, assuming that no device would want that.

**Comment by cqc-melf** [comment]
(I will put out another RC after this is merged and check that with pytket-quantinuum)

File: pytket/qir/conversion/conversion.py [hunk 3]
```diff
@@ -630,7 +632,7 @@ def condition_block() -> None:
                     [
                         pyqir.const(
                             pyqir.Type.double(self.module.module.context),
-                            float(op.params[0]),
+                            float(float(op.params[0]) * math.pi),
```
Comment (cqc-alec) [inline pytket/qir/conversion/conversion.py]: Shouldn't need the outer `float()` conversion.

File: pytket/qir/conversion/conversion.py [hunk 4]
```diff
@@ -663,16 +665,59 @@ def condition_block() -> None:
...
```
Comment (cqc-alec) [inline pytket/qir/conversion/conversion.py]: Nor here.

File: pytket/qir/conversion/conversion.py [hunk 5]
```diff
@@ -663,16 +665,59 @@ def condition_block() -> None:
...
                         ),
                         pyqir.const(
                             pyqir.Type.double(self.module.module.context),
-                            float(op.params[1]),
+                            float(float(op.params[1]) * math.pi),
```
Comment (cqc-alec) [inline pytket/qir/conversion/conversion.py]: Nor here.

File: pytket/qir/conversion/conversion.py [hunk 6]
```diff
@@ -663,16 +665,59 @@ def condition_block() -> None:
...
+                            ],
+                        ),
+                    )
+
+                self.module.builder.call(  # type: ignore
+                    self.additional_quantum_gates[OpType.TK2],
+                    [
+                        pyqir.const(
+                            pyqir.Type.double(self.module.module.context),
+                            float(float(op.params[0]) * math.pi),
```
Comment (cqc-alec) [inline pytket/qir/conversion/conversion.py]: etc

Comment (cqc-melf) [reply pytket/qir/conversion/conversion.py]: done in @cqc-melf
[remove float conversion](https://github.com/CQCL/pytket-qir/pull/43/commits/e5ae02f70fe8bb6c8705840d0a244a70a83fa78e)
[e5ae02f](https://github.com/CQCL/pytket-qir/pull/43/commits/e5ae02f70fe8bb6c8705840d0a244a70a83fa78e)

File: pytket/qir/conversion/conversion.py (hunk 5 above)
Comment (cqc-melf) [reply pytket/qir/conversion/conversion.py]: done in @cqc-melf
[remove float conversion](https://github.com/CQCL/pytket-qir/pull/43/commits/e5ae02f70fe8bb6c8705840d0a244a70a83fa78e)
[e5ae02f](https://github.com/CQCL/pytket-qir/pull/43/commits/e5ae02f70fe8bb6c8705840d0a244a70a83fa78e)

File: pytket/qir/conversion/conversion.py (hunk 4 above)
Comment (cqc-melf) [reply pytket/qir/conversion/conversion.py]: done in @cqc-melf
[remove float conversion](https://github.com/CQCL/pytket-qir/pull/43/commits/e5ae02f70fe8bb6c8705840d0a244a70a83fa78e)
[e5ae02f](https://github.com/CQCL/pytket-qir/pull/43/commits/e5ae02f70fe8bb6c8705840d0a244a70a83fa78e)

File: pytket/qir/conversion/conversion.py (hunk 3 above)
Comment (cqc-melf) [reply pytket/qir/conversion/conversion.py]: done in @cqc-melf
[remove float conversion](https://github.com/CQCL/pytket-qir/pull/43/commits/e5ae02f70fe8bb6c8705840d0a244a70a83fa78e)
[e5ae02f](https://github.com/CQCL/pytket-qir/pull/43/commits/e5ae02f70fe8bb6c8705840d0a244a70a83fa78e)

**Review by cqc-alec — APPROVED** [review]

**Pull request merged**